samples/cpp/benchmark/throughput_benchmark/main.cpp

```diff
@@ -0,0 +1,132 @@
+            ireq.wait();
+        }
+        // Benchmark for seconds_to_run seconds and at least niter iterations
+        std::chrono::seconds seconds_to_run{10};
+        size_t niter = 10;
+        std::vector<double> latencies;
+        std::mutex mutex;
+        std::condition_variable cv;
+        std::exception_ptr callback_exception;
+        struct TimedIreq {
+            ov::InferRequest& ireq;  // ref
+            std::chrono::steady_clock::time_point start;
+            bool has_start_time;
+        };
+        std::deque<TimedIreq> finished_ireqs;
+        for (ov::InferRequest& ireq : ireqs) {
+            finished_ireqs.push_back({ireq, std::chrono::steady_clock::time_point{}, false});
+        }
+        auto start = std::chrono::steady_clock::now();
+        auto time_point_to_finish = start + seconds_to_run;
+        // Once there’s a finished ireq wake up main thread.
+        // Compute and save latency for that ireq and prepare for next inference by setting up callback.
+        // Callback pushes that ireq again to finished ireqs when infrence is completed.
+        // Start asynchronous infer with updated callback
+        for (;;) {
+            std::unique_lock<std::mutex> lock(mutex);
+            while (!callback_exception && finished_ireqs.empty()) {
+                cv.wait(lock);
+            }
+            if (callback_exception) {
+                std::rethrow_exception(callback_exception);
+            }
+            if (!finished_ireqs.empty()) {
+                auto time_point = std::chrono::steady_clock::now();
+                if (time_point > time_point_to_finish && latencies.size() > niter) {
+                    break;
+                }
+                TimedIreq timedIreq = finished_ireqs.front();
+                finished_ireqs.pop_front();
+                lock.unlock();
+                ov::InferRequest& ireq = timedIreq.ireq;
+                if (timedIreq.has_start_time) {
+                    latencies.push_back(std::chrono::duration_cast<Ms>(time_point - timedIreq.start).count());
+                }
+                ireq.set_callback(
+                    [&ireq, time_point, &mutex, &finished_ireqs, &callback_exception, &cv](std::exception_ptr ex) {
+                        // Keep callback small. This improves performance for fast (tens of thousands FPS) models
+                        std::unique_lock<std::mutex> lock(mutex);
+                        {
+                            try {
+                                if (ex) {
+                                    std::rethrow_exception(ex);
+                                }
+                                finished_ireqs.push_back({ireq, time_point, true});
+                            } catch (const std::exception&) {
+                                if (!callback_exception) {
+                                    callback_exception = std::current_exception();
+                                }
+                            }
```

Comment:
I have seen this design pattern, but I am curious, why this wouldn't be better?
```suggestion
                                if (ex) {
                                    if (!callback_exception)
                                      callback_exception = ex;
                                } else {
                                  finished_ireqs.push_back({ireq, time_point, true});
                                }
```

Reply:
vector's push_back() may throw. we need to catch it. And to handle all callback's exceptions in one place we rethrow `ex` coming from callback arg.

Actually the solution isn't ideal because `std::unique_lock<std::mutex> lock(mutex);` may also throw. I think it could be solved with atomic<bool> usage. But that would also require updating other samples. So I'm leaving it for now